ホーンテッドブログの脆弱性を修正

uchihiro04 opened the pull request:
フィヨルドブートキャンプの課題になります。

ホーンテッドブログについて、以下の脆弱性を修正しました。

* 検索キーワードに`') OR secret = TRUE --`を入力すると秘密のブログが見えてしまう（SQLインジェクション）
* ブログ本文に`<script>alert("XSS")</script>`を書き込むとJavaScriptが実行されてしまう（XSS）
* 他の人のブログ記事を編集・削除できてしまう
* 他の人の秘密のブログにアクセスできてしまう
* デベロッパーツールでHTMLを編集し、`<input type="hidden" name="blog[random_eyecatch]" value="1">`を追加すると一般ユーザーでも有料機能が使えてしまう
* デベロッパツールでAPIのレスポンスを見ると、個人情報（メールアドレス）が判明してしまう

uchihiro04 committed:
SQLインジェクションを修正

SQLインジェクションに対処できていないコードについて、プレースホルダーを用いて対処しました。

uchihiro04 committed:
XSS対策を実装

XSSが発生してしまう箇所について、エスケープ処理を施しXSSが発生しないようにしました。

uchihiro04 committed:
プレースホルダーの表記を変更

プレースホルダーの表記をハッシュに変更しました。

uchihiro04 committed:
他のユーザーの記事が編集・削除できてしまう問題を修正、他のユーザーの秘密のブログを閲覧できてしまう問題を修正

URLを直打ちすると他のユーザーの記事が編集・削除できてしまう問題を修正しました。また、URLを直打ちすると他のユーザーの秘密のブログを閲覧できてしまう問題を修正しました。

uchihiro04 committed:
一般ユーザーでも有料機能が使えてしまう問題を修正

デベロッパーツールを使用すると、一般ユーザーでも有料機能が使えてしまう問題を修正しました。

uchihiro04 committed:
APIのレスポンスをのぞくと個人情報が見えてしまう問題を修正

デベロッパーツールを使用しAPIのレスポンスをのぞくと、個人情報が見えてしまう問題を修正しました。

uchihiro04 committed:
ダブルクォーテーションをシングルクォーテーションに修正

rubocopの指摘を受け、ダブルクォーテーションをシングルクォーテーションに変更しました。

uchihiro04 committed:
無効な有料機能に関するメソッドを実装

無効な有料機能に関するメソッド(invalid_paid_service?)を実装しました。

cafedomancer commented on app/helpers/blogs_helper.rb:
```diff
@@ -2,6 +2,6 @@
 module BlogsHelper
   def format_content(blog)
-    blog.content.gsub("\n", '<br>').html_safe # rubocop:disable Rails/OutputSafety
+    h(blog.content).gsub("\n", '<br>').html_safe # rubocop:disable Rails/OutputSafety
```
`rubocop:disable` がなくても RuboCop をパスするような実装にできないか考えてみてください。

uchihiro04 replied:
`html_safe`を使うとrubocopに指摘されることを初めて知りました😅
[Class: RuboCop::Cop::Rails::OutputSafety — Documentation for rubocop \(0\.41\.2\)](https://www.rubydoc.info/gems/rubocop/0.41.2/RuboCop/Cop/Rails/OutputSafety)

`html_safe`の使用をやめ、Viewで`simple_format`を使う形に変更しました。該当するコミットは[8a8d3e6](https://github.com/uchihiro04/haunted-blog/pull/1/commits/8a8d3e69bbe1a60158ad7b5083e6422dccae89e9)となります。

cafedomancer commented on app/views/api/liking_users/_user.json.jbuilder:
```diff
@@ -1,2 +1,2 @@
-json.extract! user, :id, :email, :created_at, :updated_at, :nickname, :premium
+json.extract! user, :id, :created_at, :updated_at, :nickname, :premium
```
email 以外にもフロント側に公開する必要のない情報が含まれています。情報を精査した上で不要な項目は除外してください。

uchihiro04 replied:
初回提出時は、メールアドレスが勝手に公開されてしまうのは良くないと考えたところで思考がストップしてしまいました😵

初回提出時からさらに情報を精査していきました。画面では日報に誰がLikeをしたのかだけが表示されます。それを踏まえて以下のように考えました。
* id
画面に表示されず、フロント側にLikeしたユーザーのidを表示する必要性は薄いと考え削除しました。
* created_at, :updated_at
画面に表示されず、Likeを行った時間についてフロント側が必要になる可能性は薄いと考え削除しました。
* nickname
画面ではLikeを行ったユーザーのニックネームが表示されるため、APIの返却値としてニックネームは公開されていても問題ないのではと考えそのままにしました。
* premium
画面に表示されず、フロント側にユーザーがプレミアムユーザーかどうかを公開する必要性は薄いと考え削除しました。

結論として、nicknameのみを残す形に変更しました。該当するコミットは[763f998](https://github.com/uchihiro04/haunted-blog/pull/1/commits/763f998a7ec655ef0498a6a28d31ebbc9b3994fb)となります。

cafedomancer replied:
変更内容としては OK です。「画面に表示されない」というのがフロントエンド側の実装を見た上での判断なのか、単に画面を確認したのかのどちらなのかがちょっと分からなかったのですが、この変更内容でよいと確実に判断するためにはフロントエンド側の実装を確認する必要があります。フロントエンド側の実装をもし確認していないのであれば、確認してみた上でこの変更内容でよいことを改めて判断してみてください。

uchihiro04 replied:
画面での確認に留まっておりましたため、実装について確認いたしました。

ボブのブログを開いた場合、以下の画面になります。
<img width="823" alt="image" src="https://user-images.githubusercontent.com/77523896/213970449-fb111b81-f6ea-43c8-897d-f263489fb6be.png">

今回は`liking_users_path` が `/api/blogs/3/liking_users.json` であることがわかりました。
`/api/blogs/3/liking_users.json` は、`app/views/api/liking_users/_user.json.jbuilder` で設定されている、Like したユーザーの情報が入った JSON ファイルです。
https://github.com/uchihiro04/haunted-blog/blob/751115e4a26e2606f8478f806c5bcf449e0ded4e/app/views/blogs/_blog.html.erb#L18-L22

`liking_users_path` は `app/javascript/vuejs/liking.js` にて `likingUsersPath` という変数に代入されます。
https://github.com/uchihiro04/haunted-blog/blob/751115e4a26e2606f8478f806c5bcf449e0ded4e/app/javascript/vuejs/liking.js#L9

変数`likingUsersPath`は`app/javascript/vuejs/LikingApp.js`内の`fetchLikingUsers`メソッドで使用されます。json ファイルの Like したユーザーに関する情報が `users` に代入されます。
https://github.com/uchihiro04/haunted-blog/blob/751115e4a26e2606f8478f806c5bcf449e0ded4e/app/javascript/vuejs/LikingApp.js#L42-L45

その後 `users` は繰り返し処理でLikeしたユーザーのニックネームを表示するために使用されます。
https://github.com/uchihiro04/haunted-blog/blob/751115e4a26e2606f8478f806c5bcf449e0ded4e/app/javascript/vuejs/LikingApp.js#L13

ユーザー側へは Like したユーザーの情報のうち `nickname` しか表示しておらず、その他の情報は表示していません。そのため、APIの返却値も`nickname` のみを公開する形とすることを、改めて判断いたしました。

cafedomancer replied:
よさそうです 🙆 ちなみに JSON 形式のデータではあるのですが JSON **ファイル**ではありません。

uchihiro04 replied:
コメントありがとうございます！
JSONファイルではない旨、承知しました。今後説明を行う際は気をつけたいと思います。

cafedomancer commented on app/controllers/blogs_controller.rb:
```diff
@@ -43,8 +51,12 @@ def destroy
 
-  def set_blog
-    @blog = Blog.find(params[:id])
+  def correct_user
```
`correct_user` と言いつつここで検索した `@blog` を各アクションで使用しているので、メソッド名と実態とが合っていない気がしました。メソッド名を見直してみてください。

uchihiro04 replied:
考え直してみると、現在`correct_user`メソッドで行っていることは、検索したブログをインスタンス変数`@blog`に定義しているだけですね😵

インスタンス変数`@blog`をセットして他のアクションで使う形にしているので、`set_blog`にメソッド名を変更しました。該当するコミットは[323801c](https://github.com/uchihiro04/haunted-blog/pull/1/commits/323801c770a1cfd9f4c57963c4a8d4db30821fd3)となります。

cafedomancer commented on app/controllers/blogs_controller.rb:
```diff
@@ -21,6 +27,7 @@ def create
     @blog = current_user.blogs.new(blog_params)
 
     if @blog.save
+      @blog.update(random_eyecatch: false) if invalid_paid_service?
```
この実装だと以下の 2 つの問題が生じます。

- L29 の `@blog.save` と L30 の `@blog.update` は別々のトランザクションで実行されるため、L29 の実行には成功したが L30 の実行には失敗し、データの不整合が発生する、といった状況が起こり得る
- L30 の `@blog.update` で `blogs.updated_at` が余計に更新される

上記 2 点の問題をクリアするような実装を考えてみてください。update アクションについても同様です。

uchihiro04 replied:
いただいたレビューを受け、ストロングパラメーターを使って許可するキーを変更する形に修正しました。
下記の通り、「別々のトランザクションで実行される」「`blogs.updated_at`が余計に更新される」問題は発生しませんでした。該当するコミットは[b5ca540](https://github.com/uchihiro04/haunted-blog/pull/1/commits/b5ca5402fe9464aab23ed465b05df59ed06e29e8)となります。

createメソッド
```sql
Started POST "/blogs" for 127.0.0.1 at 2023-01-21 13:20:44 +0900
Processing by BlogsController#create as TURBO_STREAM
  Parameters: {"authenticity_token"=>"[FILTERED]", "blog"=>{"title"=>"ボブ", "content"=>"ボブです", "secret"=>"[FILTERED]", "random_eyecatch"=>"1"}, "commit"=>"登録する"}
  User Load (0.2ms)  SELECT "users".* FROM "users" WHERE "users"."id" = ? ORDER BY "users"."id" ASC LIMIT ?  [["id", 2], ["LIMIT", 1]]
Unpermitted parameter: :random_eyecatch. Context: { controller: BlogsController, action: create, request: #<ActionDispatch::Request:0x00000001082fec30>, params: {"authenticity_token"=>"[FILTERED]", "blog"=>{"title"=>"ボブ", "content"=>"ボブです", "secret"=>"[FILTERED]", "random_eyecatch"=>"1"}, "commit"=>"登録する", "controller"=>"blogs", "action"=>"create"} }
  TRANSACTION (0.1ms)  begin transaction
  ↳ app/controllers/blogs_controller.rb:32:in `create'
  Blog Create (0.9ms)  INSERT INTO "blogs" ("title", "content", "created_at", "updated_at", "secret", "user_id", "random_eyecatch") VALUES (?, ?, ?, ?, ?, ?, ?)  [["title", "ボブ"], ["content", "ボブです"], ["created_at", "2023-01-21 04:20:44.199787"], ["updated_at", "2023-01-21 04:20:44.199787"], ["secret", "[FILTERED]"], ["user_id", 2], ["random_eyecatch", 0]]
  ↳ app/controllers/blogs_controller.rb:32:in `create'
  TRANSACTION (0.8ms)  commit transaction
  ↳ app/controllers/blogs_controller.rb:32:in `create'
Redirected to http://127.0.0.1:3000/blogs/10
Completed 302 Found in 19ms (ActiveRecord: 2.0ms | Allocations: 7694)
```

updateメソッド
```sql
Started PATCH "/blogs/10" for 127.0.0.1 at 2023-01-22 11:07:39 +0900
Processing by BlogsController#update as TURBO_STREAM
  Parameters: {"authenticity_token"=>"[FILTERED]", "blog"=>{"title"=>"ボブ", "content"=>"ボブです。こんにちは。", "secret"=>"[FILTERED]"}, "commit"=>"更新する", "id"=>"10"}
  User Load (0.1ms)  SELECT "users".* FROM "users" WHERE "users"."id" = ? ORDER BY "users"."id" ASC LIMIT ?  [["id", 2], ["LIMIT", 1]]
  Blog Load (0.0ms)  SELECT "blogs".* FROM "blogs" WHERE "blogs"."user_id" = ? AND "blogs"."id" = ? LIMIT ?  [["user_id", 2], ["id", 10], ["LIMIT", 1]]
  ↳ app/controllers/blogs_controller.rb:53:in `set_blog'
  TRANSACTION (0.1ms)  begin transaction
  ↳ app/controllers/blogs_controller.rb:37:in `update'
  Blog Update (1.0ms)  UPDATE "blogs" SET "content" = ?, "updated_at" = ? WHERE "blogs"."id" = ?  [["content", "ボブです。こんにちは。"], ["updated_at", "2023-01-22 02:07:39.869652"], ["id", 10]]
  ↳ app/controllers/blogs_controller.rb:37:in `update'
  TRANSACTION (0.8ms)  commit transaction
  ↳ app/controllers/blogs_controller.rb:37:in `update'
Redirected to http://127.0.0.1:3000/blogs/10
Completed 302 Found in 8ms (ActiveRecord: 2.0ms | Allocations: 5623)
```

cafedomancer commented on app/controllers/blogs_controller.rb:
```diff
@@ -3,13 +3,19 @@
-  def show; end
+  def show
+    @blog = if current_user.nil?
+              Blog.where(secret: false).find(params[:id])
```
`where(secret: false)` の代わりに既存の実装を使い回せないか考えてみてください。

uchihiro04 replied:
`app/models/blog.rb`に`published`メソッドが用意されていることを見落としていました😵

`where(secret: false) `の代わりに`published`を使用する形に変更しました。該当するコミットは[ccdd02b](https://github.com/uchihiro04/haunted-blog/pull/1/commits/ccdd02bd12e1be0883911ab389e87a284e5922bf)となります。

cafedomancer replied:
`where` メソッドにキーワード引数の形で引数を渡すの自体はよいので、`published` スコープの実装をそのように変えてもらってもよいです。お好みでどうぞ。

uchihiro04 replied:
コメントありがとうございます🙇‍♂️
今回につきましては、既存の実装をそのまま使用する（キーワード引数を使わない）形をとることにしたいと思います。

cafedomancer commented on app/controllers/blogs_controller.rb:
```diff
@@ -3,13 +3,19 @@
+    @blog = if current_user.nil?
+              Blog.where(secret: false).find(params[:id])
+            else
+              Blog.find(params[:id]).secret? ? current_user.blogs.find(params[:id]) : Blog.find(params[:id])
```
この実装だと条件演算子の評価前後に SQL クエリが 2 回発行されませんか？ うまくやれば 1 回のクエリで検索できるはずなのでチャレンジしてみてください。

uchihiro04 replied:
ペアプロの後、「ログインしているユーザー自身のブログ記事か、シークレットモードではないブログ記事は閲覧できる」形にコードの修正を試みました。
現在のコードは以下の通りです。
```ruby
# app/controllers/blogs_controller.rb
# 該当するコードのみ抜粋
def show
  @blog = if Blog.find(params[:id]).owned_by?(current_user)
            Blog.find(params[:id])
          else
            Blog.published.find(params[:id])
          end
end
```

しかし、ログインしているユーザー自身のものではないブログ記事を閲覧する際、1回のクエリで検索する形にできませんでした。

（例）BobのアカウントでAliceのブログ記事を閲覧する
Blogについて、以下の2回クエリが発行されています。
* 条件を読み込むとき
`Blog Load (0.1ms)  SELECT "blogs".* FROM "blogs" WHERE "blogs"."id" = ? LIMIT ?  [["id", 7], ["LIMIT", 1]]
  ↳ app/controllers/blogs_controller.rb:13:in 'show'`
* 条件に応じた結果を表示するとき
`Blog Load (0.1ms)  SELECT "blogs".* FROM "blogs" WHERE (secret = FALSE) AND "blogs"."id" = ? LIMIT ?  [["id", 7], ["LIMIT", 1]]`

実際のログ
```sql
Started GET "/blogs/7" for 127.0.0.1 at 2023-01-22 11:17:50 +0900
Processing by BlogsController#show as HTML
  Parameters: {"id"=>"7"}
  Blog Load (0.1ms)  SELECT "blogs".* FROM "blogs" WHERE "blogs"."id" = ? LIMIT ?  [["id", 7], ["LIMIT", 1]]
  ↳ app/controllers/blogs_controller.rb:13:in `show'
  User Load (0.0ms)  SELECT "users".* FROM "users" WHERE "users"."id" = ? ORDER BY "users"."id" ASC LIMIT ?  [["id", 2], ["LIMIT", 1]]
  ↳ app/controllers/blogs_controller.rb:13:in `show'
  User Load (0.0ms)  SELECT "users".* FROM "users" WHERE "users"."id" = ? LIMIT ?  [["id", 1], ["LIMIT", 1]]
  ↳ app/models/blog.rb:19:in `owned_by?'
  Blog Load (0.1ms)  SELECT "blogs".* FROM "blogs" WHERE (secret = FALSE) AND "blogs"."id" = ? LIMIT ?  [["id", 7], ["LIMIT", 1]]
  ↳ app/controllers/blogs_controller.rb:16:in `show'
  Rendering layout layouts/application.html.erb
  Rendering blogs/show.html.erb within layouts/application
  CACHE User Load (0.0ms)  SELECT "users".* FROM "users" WHERE "users"."id" = ? LIMIT ?  [["id", 1], ["LIMIT", 1]]
  ↳ app/views/blogs/_blog.html.erb:4
  Rendered blogs/_blog.html.erb (Duration: 1.1ms | Allocations: 1021)
  Rendered blogs/show.html.erb within layouts/application (Duration: 1.5ms | Allocations: 1140)
  Rendered layouts/_navbar.html.erb (Duration: 0.3ms | Allocations: 100)
  Rendered layouts/_flash.html.erb (Duration: 0.0ms | Allocations: 29)
  Rendered layout layouts/application.html.erb (Duration: 4.7ms | Allocations: 3792)
Completed 200 OK in 8ms (Views: 5.0ms | ActiveRecord: 0.2ms | Allocations: 6791)
```

現在showメソッドにつきまして、1 回のクエリで検索するためのコードを書けずに詰まっている状態です。
お手数をおかけしますが、アドバイスをいただけますと幸いです🙇‍♂️

cafedomancer replied:
まず最初に、検索したいデータはどのような SQL で検索できるのかを考えてください。それができたら、その SQL を発行するような Active Record のコードを考えてください。Ruby のコードから考え始めると、Active Record パターン (1 レコードが 1 オブジェクトにマッピングされるようなパターン) が採用されている都合で、どうしても 1 件 1 件のレコード (オブジェクト) に着目せざるを得なくなってしまい、手に入れたいデータの集合はどのようなものかという観点が抜け落ちてしまいます。1 度のクエリで検索を済ませるには、今回得たいデータの集合がどのようなクエリで得られるかということから考え始める必要があります。

uchihiro04 replied:
アドバイスありがとうございます🙇‍♂️
SQLから考えることで、手に入れたいデータの集合に着目することができました。

showメソッド実行時に、1回のクエリでブログを検索できる形に修正しました。該当するコミットは[c93cc67](https://github.com/uchihiro04/haunted-blog/pull/1/commits/c93cc677418e77693fc5b934e354f640dd23fb79)となります。
`app/controllers/blogs_controller.rb` の `show` メソッドを実行した際のログは以下の通りです。

* 閲覧可能なブログにアクセスした場合
```sql
Started GET "/blogs/1" for 127.0.0.1 at 2023-01-23 17:25:12 +0900
Processing by BlogsController#show as HTML
  Parameters: {"id"=>"1"}
  User Load (0.2ms)  SELECT "users".* FROM "users" WHERE "users"."id" = ? ORDER BY "users"."id" ASC LIMIT ?  [["id", 2], ["LIMIT", 1]]
  ↳ app/controllers/blogs_controller.rb:13:in `show'
  Blog Load (0.2ms)  SELECT "blogs".* FROM "blogs" WHERE ("blogs"."user_id" = ? OR secret = FALSE) AND "blogs"."id" = ? LIMIT ?  [["user_id", 2], ["id", 1], ["LIMIT", 1]]
  ↳ app/controllers/blogs_controller.rb:13:in `show'
  Rendering layout layouts/application.html.erb
  Rendering blogs/show.html.erb within layouts/application
  User Load (0.1ms)  SELECT "users".* FROM "users" WHERE "users"."id" = ? LIMIT ?  [["id", 1], ["LIMIT", 1]]
  ↳ app/views/blogs/_blog.html.erb:4
  Rendered blogs/_blog.html.erb (Duration: 2.9ms | Allocations: 1119)
  Rendered blogs/show.html.erb within layouts/application (Duration: 3.2ms | Allocations: 1251)
  Rendered layouts/_navbar.html.erb (Duration: 0.1ms | Allocations: 103)
  Rendered layouts/_flash.html.erb (Duration: 0.0ms | Allocations: 35)
  Rendered layout layouts/application.html.erb (Duration: 7.2ms | Allocations: 3949)
Completed 200 OK in 13ms (Views: 7.6ms | ActiveRecord: 0.4ms | Allocations: 6185)
```

* 閲覧不可能なブログにアクセスした場合
```sql
Started GET "/blogs/2" for 127.0.0.1 at 2023-01-23 17:26:51 +0900
Processing by BlogsController#show as HTML
  Parameters: {"id"=>"2"}
  User Load (0.9ms)  SELECT "users".* FROM "users" WHERE "users"."id" = ? ORDER BY "users"."id" ASC LIMIT ?  [["id", 2], ["LIMIT", 1]]
  ↳ app/controllers/blogs_controller.rb:13:in `show'
  Blog Load (0.7ms)  SELECT "blogs".* FROM "blogs" WHERE ("blogs"."user_id" = ? OR secret = FALSE) AND "blogs"."id" = ? LIMIT ?  [["user_id", 2], ["id", 2], ["LIMIT", 1]]
  ↳ app/controllers/blogs_controller.rb:13:in `show'
Completed 404 Not Found in 27ms (ActiveRecord: 1.7ms | Allocations: 2354)
```

cafedomancer commented on app/controllers/blogs_controller.rb:
```diff
@@ -3,13 +3,19 @@
 
-  def show; end
+  def show
+    @blog = if current_user.nil?
```
`current_user` が `nil` である (= ログインしていない) という否定の条件を先に書くより、肯定の条件を先に書くほうが読むときの負荷が少ないかなーと思います。

uchihiro04 replied:
確かに肯定の条件→否定の条件と書いた方が、読み下すときに自然ですね🙂

肯定の条件を先に書く形に修正しました。該当するコミットは[180455a](https://github.com/uchihiro04/haunted-blog/pull/1/commits/180455aad01f0b017d2350ac539a9929e88c3c55)です。

uchihiro04 committed:
条件分岐について、肯定条件を先に書くように変更

レビューを受け、肯定条件を先に書くように変更しました。

uchihiro04 committed:
APIの返却値を変更

APIの返却値について、ニックネームのみに変更しました。

uchihiro04 committed:
メソッド名の変更

レビューを受け、メソッド名をcorrect_userメソッドからset_blogメソッドに変更しました。

uchihiro04 committed:
rubocop:disable がなくても RuboCop をパスするよう修正

レビューを受け、rubocop:disable がなくても RuboCop をパスするように修正しました。

uchihiro04 committed:
無料会員でも有料機能が使えてしまう問題の修正

レビューを受け、ストロングパラメーターを使って無料会員でも有料機能が使えてしまう問題を修正しました。トランザクションが2回起こる・更新日時が余計に更新される問題も解消されました。

uchihiro04 committed:
条件分岐について、present?を使った形に変更

showメソッドの条件分岐について、present?を使った形に変更しました。

uchihiro04 committed:
whereメソッドをスコープに変更

シークレットモードに関する条件分岐について、whereメソッドからスコープを使う形に修正しました。

uchihiro04 committed:
showメソッドの条件分岐を変更

showメソッドの条件分岐について、owned_by?メソッドを使う形に変更しました。

uchihiro04 committed:
showメソッドのインデントを修正

rubocopからの指摘を受け、showメソッドのインデントを修正しました。

cafedomancer commented on app/controllers/blogs_controller.rb:
```diff
@@ -44,10 +50,18 @@ def destroy
+    @blog = current_user.blogs.find(params[:id])
+  end
+
+  def invalid_paid_service?
```
このメソッドはまだどこかで使っていますか？

cafedomancer commented on app/views/blogs/_blog.html.erb:
```diff
@@ -13,7 +13,7 @@
     <% end %>
     <div class="blog-content">
-      <%= format_content(blog) %>
+      <%= simple_format(format_content(blog)) %>
```
`blog` の `content` を表示するために `simple_format` を実行しなければらないのであれば、`simple_format` も `format_content` の処理の一部ではないかなーと思いました。そうしておけば、他のところで `blog` の `content` を表示することになっても、`format_content` を呼ぶだけで済むようになると思います。

uchihiro04 replied:
確かに `simple_format`を `format_content` の処理の一部にしたほうが、再利用しやすくなりますね😅

`simple_format` を `format_content` の処理の一部にする形に修正しました。該当するコミットは[785c550](https://github.com/uchihiro04/haunted-blog/pull/1/commits/785c55008bded7e85a0dfee30907e45e51f90cb3)となります。

cafedomancer commented on app/controllers/blogs_controller.rb:
```diff
@@ -44,10 +50,18 @@ def destroy
+    @blog = current_user.blogs.find(params[:id])
+  end
+
+  def invalid_paid_service?
```
このメソッドはまだどこかで使っていますか？

uchihiro04 replied:
ストロングパラメーターを使用した修正（該当コミット[b5ca540](https://github.com/uchihiro04/haunted-blog/pull/1/commits/b5ca5402fe9464aab23ed465b05df59ed06e29e8)）を行った際に、消し忘れておりました😵

使用していないメソッドのため、削除しました。該当するコミットは[052ad2e](https://github.com/uchihiro04/haunted-blog/pull/1/commits/052ad2e2e7c4bdbd3f008f7202a4cea0eb21be6e)となります。

cafedomancer commented on app/helpers/blogs_helper.rb:
```diff
@@ -2,6 +2,6 @@
 module BlogsHelper
   def format_content(blog)
-    blog.content.gsub("\n", '<br>').html_safe # rubocop:disable Rails/OutputSafety
+    h(blog.content).gsub("\n", '<br>')
```
`gsub` の処理は `simple_format` の導入で不要になっている気がします。`simple_format` のドキュメントを確認してみてください。https://api.rubyonrails.org/classes/ActionView/Helpers/TextHelper.html#method-i-simple_format

uchihiro04 replied:
`simple_format` のドキュメントを確認いたしました。
[ActionView::Helpers::TextHelper](https://api.rubyonrails.org/classes/ActionView/Helpers/TextHelper.html#method-i-simple_format)

`simple_format` のみの場合、改行が1回の場合はテキストエリア内で入力したものと同じものが表示されます。しかし、2回以上連続して改行を行うと、段落とみなされ`<p>`タグで囲まれます。
そのため、2回以上連続して改行しても改行は1つ分しか反映されず、テキストエリア内で入力したものと差異が生じます。
![23012301](https://user-images.githubusercontent.com/77523896/213969316-b1f62710-a34a-4e0e-b5d9-8a98f7a06ac0.gif)

対して、`gsub`を使用した場合は2つ以上の連続した改行でも段落とみなされないため、そのまま改行が反映されます。
![23012302](https://user-images.githubusercontent.com/77523896/213969346-3720b818-b9f4-4422-8c4a-7bc4e4eb18b9.gif)

今回は、`textarea`タグで入力した内容をそのまま表示した方が良いのではないかと個人的には考えており、`gsub` を使わない形への変更を保留した状態にしております。

cafedomancer replied:
入力したとおりに表示するために現在の実装を使用するのは OK です 🙆 だとすると `simple_format` は何のために使っているんでしたっけ？

uchihiro04 replied:
`simple_format`はテキストをHTMLに変換したものを返します。

`simple_format`を使わずに、`format_content`メソッドの中身を`h(blog.content).gsub("\n", '<br>')`とすると、`<script>`タグの`<>`をはじめとした記号はエスケープ処理された形で表示されます。また、`gsub`メソッドで置き換えられた`<br>`タグはそのまま文字列として表示されます。
![image.png](https://bootcamp.fjord.jp/rails/active_storage/blobs/redirect/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBMURvQWc9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--1349e71319b056d5ffeca059d7490abb0bfcfaa8/image.png)

対して、`format_content`メソッドの中身を`simple_format(h(blog.content).gsub("\n", '<br>'))`とした場合、`<script>`タグの`<>`をはじめとした記号は文字となり、`gsub`メソッドで置き換えられた`<br>`タグは改行として表示されます。
![image.png](https://bootcamp.fjord.jp/rails/active_storage/blobs/redirect/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBMGJvQWc9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--82d06ea459601cf683cceba6f2b2899518461913/image.png)

今回`simple_format`を使用したのは、`<script>`タグの`<>`をはじめとした記号を文字として、`<br>`タグを改行として表示するために使用しています。

もし、いただいた質問の意図と違う回答になっておりましたら、ご指摘いただけますと幸いです🙇‍♂️

cafedomancer replied:
あーなるほど、わかりました！ こちらで大丈夫ですー 👌

uchihiro04 committed:
不要なメソッド（invalid_paid_service?）を削除

uchihiro04 committed:
simple_formatをformat_contentの処理の一部にする形に修正

再利用のしやすさを考え、simple_formatをformat_contentの処理の一部にする形に修正しました。

uchihiro04 committed:
showメソッド実行時のBlogのクエリ発行を2回から1回に修正

cafedomancer commented on app/controllers/blogs_controller.rb:
```diff
@@ -3,13 +3,15 @@
 
-  def show; end
+  def show
+    @blog = Blog.where(user_id: current_user).or(Blog.published).find(params[:id])
```
特定のユーザーに紐づくデータを参照する場合は、ユーザーの関連からデータを引くようにしてください。つまり、`Blog` ではなく `current_user` からデータを引く形になります。理由は、ユーザーの絞り込み条件を書き忘れて情報が漏洩することを防ぐためです。

uchihiro04 replied:
ご指摘ありがとうございます🙇‍♂️
特定のユーザーに紐づくデータを参照する場合、ユーザーの関連からデータを参照する形に変更しました。該当するコミットは[ebdb27d](https://github.com/uchihiro04/haunted-blog/pull/1/commits/ebdb27d24dfd1b8c2da78e117d5a85a7c31af3ff)となります。

cafedomancer replied:
おっとすみません、これは `current_user` が `nil` になるケースをうまく使っていたのですね。であれば、もとの書き方でよいです 🙇 ちなみにオブジェクトを渡す場合は `user_id` ではなく `user` (Active Record で定義しているアソシエーションの名前) で指定できるので、そちらを使うとよいと思います。

uchihiro04 replied:
承知しました🙂
一度コードを元に戻した後、`where`メソッド内のキーを`user_id`から`user`に変更しました。該当するコミットは[133da9f](https://github.com/uchihiro04/haunted-blog/pull/1/commits/133da9fd4620be28f9aeb74eea58bc1027ceb761)となります。

uchihiro04 committed:
特定のユーザーに紐づくデータを参照する場合、ユーザーの関連からデータを引く形に修正

uchihiro04 committed:
条件分岐を削除し、1行で記述する形に再変更

1行で記述する形に再変更した際、キーをuser_idからuserに変更しました。
